### PR TITLE
Add the abstract GenerativeEnv

### PR DESCRIPTION
**Pull request by hongkai-dai**
Also modify the API for DynamicsModel, that step() function takes state/action and returns the next state.

- - -
This change is [Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17)

### CODE REVIEW

**Review by hongkai-dai — COMMENTED** [review]
+@blake-wulfe-tri for review.

Sorry for the delay, I got distracted by some Drake PRs.

> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-:-NOs-Bv94-kmnJGL4_q3:b-jcativ)* status: 0 of 8 files reviewed, all discussions resolved (waiting on @blake-wulfe-tri)

**Review by blake-wulfe-tri — COMMENTED** [review]
No problem, thanks for putting this together! :lgtm: 

> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-:-NOuRiTT5FZExW1UtNwk:b-ysxd7z)* status: 0 of 8 files reviewed, 5 unresolved discussions (waiting on @hongkai-dai)

___
*[`mani_skill2/dynamics/base.py` line 6 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-NOuQQ0H4TYF-oeCCkPc:-NOuQQ0H4TYF-oeCCkPd:bdu1goo) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/1a985edde23c45f227d2dee861f5666cfac54905/mani_skill2/dynamics/base.py#L6)):*
> ```Python
> import torch
> 
> StateType = TypeVar("StateType")
> ```

btw not to slow things down, but I can imagine that subclasses of `DynamicsModel` might want to actually specify what the type is for `StateType`. If that's the case I think though am not positive that using `Generic` as as base class for `DynamicsModel` might make sense: https://docs.python.org/3/library/typing.html#user-defined-generic-types

Also right now `DynamicsModel` inherits `nn.Module`, but given the `StateType` and stuff I'm not sure that makes sense
___
*[`mani_skill2/dynamics/base.py` line 31 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-NOuQyt96UC3XMJ_83ps:-NOuQyt96UC3XMJ_83pt:b-94yu8h) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/1a985edde23c45f227d2dee861f5666cfac54905/mani_skill2/dynamics/base.py#L31)):*
> ```Python
>         """Resets any internal state; called at the beginning of an episode."""
> 
>     def observation(self, state: StateType)->ObsType:
> ```

btw I think this isn't formatted according to the black standards
___
*[`mani_skill2/dynamics/reward.py` line 21 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-NOuR6xM5nuDYX4DDUNY:-NOuR6xM5nuDYX4DDUNZ:b-fy367j) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/1a985edde23c45f227d2dee861f5666cfac54905/mani_skill2/dynamics/reward.py#L21)):*
> ```Python
> 
>     @abc.abstractmethod
>     def step(
> ```

btw I believe that for POMDPs the reward might in theory need to depend on the next state, but for MDPs I think you can just use r(s,a). Given that does it make sense to leave out `next_state` and `next_obs` in order to simplify things?
___
*[`mani_skill2/tests/dynamics/test_generative_env.py` line 38 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-NOuRZNO1_Y5cm9jLr0k:-NOuRZNO1_Y5cm9jLr0l:bth3f1u) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/1a985edde23c45f227d2dee861f5666cfac54905/mani_skill2/tests/dynamics/test_generative_env.py#L38)):*
> ```Python
> class MockGenerativeEnv(GenerativeEnv):
>     dynamics_model: NetworkDynamicsModel
>     reward_model: MockRewardModel
> ```

btw if the base class already specifies these types do you have to do it again in the subclass? 
___
*[`mani_skill2/tests/dynamics/test_generative_env.py` line 67 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-NOuRffEAy-A-mEnczOK:-NOuRffEAy-A-mEnczOL:b-rnsp5z) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/1a985edde23c45f227d2dee861f5666cfac54905/mani_skill2/tests/dynamics/test_generative_env.py#L67)):*
> ```Python
> 
> class TestGenerativeEnv:
>     @pytest.mark.parametrize("state_size", (1, 3))
> ```

btw nice use of pytest!

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17)* status: 0 of 8 files reviewed, 4 unresolved discussions (waiting on @blake-wulfe-tri)

___
*[`mani_skill2/dynamics/base.py` line 6 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-NOuQQ0H4TYF-oeCCkPc:-NOuU4qp1kv6f6h9ZDzw:bzhkgbu) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/1a985edde23c45f227d2dee861f5666cfac54905/mani_skill2/dynamics/base.py#L6)):*
<details><summary><i>Previously, blake-wulfe-tri wrote…</i></summary><blockquote>

btw not to slow things down, but I can imagine that subclasses of `DynamicsModel` might want to actually specify what the type is for `StateType`. If that's the case I think though am not positive that using `Generic` as as base class for `DynamicsModel` might make sense: https://docs.python.org/3/library/typing.html#user-defined-generic-types

Also right now `DynamicsModel` inherits `nn.Module`, but given the `StateType` and stuff I'm not sure that makes sense
</blockquote></details>

For my own educational purpose, just to make sure that I understand the issue, if we use generic `StateType` for the base DynamicsModel, for the child class, I think we can still use a concrete type? For example in `NetworkDynamicsModel`, its step function annotates the type as `torch.Tensor`



Is this a common practice in python programming (use generic type for the base class, and concrete type for the child class)?

I think it makes sense to change all of these types (state / action / observation) to torch.Tensor, as `DynamicsModel` is a torch.nn.Module any way. This also mean that if we use the Maniskill simulator as the dynamics model (which operates on numpy array instead of torch Tensor), we will need to convert that torch tensor to numpy arrays. Is that a common practice in machine learning (to convert back and forth between torch.Tensor and numpy array)?
___
*[`mani_skill2/dynamics/base.py` line 31 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-NOuQyt96UC3XMJ_83ps:-NOvEUQh-xO3ltnrtmB6:b-896fix) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/1a985edde23c45f227d2dee861f5666cfac54905/mani_skill2/dynamics/base.py#L31)):*
<details><summary><i>Previously, blake-wulfe-tri wrote…</i></summary><blockquote>

btw I think this isn't formatted according to the black standards
</blockquote></details>

Done.
___
*[`mani_skill2/dynamics/reward.py` line 21 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-NOuR6xM5nuDYX4DDUNY:-NOvEa8s0K3HtrNd0Fyp:b7nsfax) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/1a985edde23c45f227d2dee861f5666cfac54905/mani_skill2/dynamics/reward.py#L21)):*
<details><summary><i>Previously, blake-wulfe-tri wrote…</i></summary><blockquote>

btw I believe that for POMDPs the reward might in theory need to depend on the next state, but for MDPs I think you can just use r(s,a). Given that does it make sense to leave out `next_state` and `next_obs` in order to simplify things?
</blockquote></details>

I agree with MDP we can always compute the next state from the current state and the action, hence putting `next_state` and `next_obs` are not strictly necessary. The reason I leave them in the input arguments is to save computation, in case the reward would depend on both the current and the next state/observation, and the next_state has already been computed from the step() function in the dynamics model.

Having said that, I cannot think of a concrete model that the reward function would depend on the next observation, especially for the dough project. Hence I removed these two arguments.
___
*[`mani_skill2/tests/dynamics/test_generative_env.py` line 38 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-NOuRZNO1_Y5cm9jLr0k:-NOv_O3J4OjLw1DNo12m:b-htn2cg) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/1a985edde23c45f227d2dee861f5666cfac54905/mani_skill2/tests/dynamics/test_generative_env.py#L38)):*
<details><summary><i>Previously, blake-wulfe-tri wrote…</i></summary><blockquote>

btw if the base class already specifies these types do you have to do it again in the subclass? 
</blockquote></details>

Ah the base class specified them as the base `DynamicsModel` and `GoalBasedRewardModel`. I was trying to be more specific here.
___
*[`mani_skill2/tests/dynamics/test_generative_env.py` line 67 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-NOuRffEAy-A-mEnczOK:-NOv_XCO7Oo8G971E8ZE:bwau1he) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/1a985edde23c45f227d2dee861f5666cfac54905/mani_skill2/tests/dynamics/test_generative_env.py#L67)):*
<details><summary><i>Previously, blake-wulfe-tri wrote…</i></summary><blockquote>

btw nice use of pytest!
</blockquote></details>

Thanks for introducing me this tool!

**Review by blake-wulfe-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17)* status: 0 of 8 files reviewed, 1 unresolved discussion (waiting on @hongkai-dai)

___
*[`mani_skill2/dynamics/base.py` line 6 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-NOuQQ0H4TYF-oeCCkPc:-NOvc9BuC-sjb-iVqyYh:b-eia0xt) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/1a985edde23c45f227d2dee861f5666cfac54905/mani_skill2/dynamics/base.py#L6)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

For my own educational purpose, just to make sure that I understand the issue, if we use generic `StateType` for the base DynamicsModel, for the child class, I think we can still use a concrete type? For example in `NetworkDynamicsModel`, its step function annotates the type as `torch.Tensor`



Is this a common practice in python programming (use generic type for the base class, and concrete type for the child class)?

I think it makes sense to change all of these types (state / action / observation) to torch.Tensor, as `DynamicsModel` is a torch.nn.Module any way. This also mean that if we use the Maniskill simulator as the dynamics model (which operates on numpy array instead of torch Tensor), we will need to convert that torch tensor to numpy arrays. Is that a common practice in machine learning (to convert back and forth between torch.Tensor and numpy array)?

</blockquote></details>

Ok looking into it a bit more I think I'm wrong and using `Generic` doesn't make sense. 
Answering your first question: I think doing this explicit typing with torch.Tensor is the right way to go
The second question: I don't know why we would need to convert to torch.Tensor when using the gt simulator. One reason you might be saying that is if we use `NetworkDynamicsModel`, and if that's the case my response is that we should just use a different concrete class (e.g., `RollingDynamicsModel`). 
___
*[`mani_skill2/dynamics/base.py` line 31 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-NOuQyt96UC3XMJ_83ps:-NOvctcd3Sw1X1EG5X8U:b-zcjvte) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/1a985edde23c45f227d2dee861f5666cfac54905/mani_skill2/dynamics/base.py#L31)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks
___
*[`mani_skill2/dynamics/reward.py` line 21 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-NOuR6xM5nuDYX4DDUNY:-NOvd2lb7cQ1v7L-hrWB:bdnes1y) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/1a985edde23c45f227d2dee861f5666cfac54905/mani_skill2/dynamics/reward.py#L21)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I agree with MDP we can always compute the next state from the current state and the action, hence putting `next_state` and `next_obs` are not strictly necessary. The reason I leave them in the input arguments is to save computation, in case the reward would depend on both the current and the next state/observation, and the next_state has already been computed from the step() function in the dynamics model.

Having said that, I cannot think of a concrete model that the reward function would depend on the next observation, especially for the dough project. Hence I removed these two arguments.
</blockquote></details>

Sounds good. 
___
*[`mani_skill2/tests/dynamics/test_generative_env.py` line 38 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-NOuRZNO1_Y5cm9jLr0k:-NOvdEUU5T7tj06WnbsA:bdsvfmk) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/1a985edde23c45f227d2dee861f5666cfac54905/mani_skill2/tests/dynamics/test_generative_env.py#L38)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Ah the base class specified them as the base `DynamicsModel` and `GoalBasedRewardModel`. I was trying to be more specific here.
</blockquote></details>

got it thanks!

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17)* status: 0 of 8 files reviewed, all discussions resolved (waiting on @blake-wulfe-tri)

___
*[`mani_skill2/dynamics/base.py` line 6 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-NOuQQ0H4TYF-oeCCkPc:-NOvfBnG-Z3A85gKZJQV:bkfqiza) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/1a985edde23c45f227d2dee861f5666cfac54905/mani_skill2/dynamics/base.py#L6)):*
Done. I changed everything to torch.Tensor

> I don't know why we would need to convert to torch.Tensor when using the gt simulator

I mean now the base class `DynamicsModel` assumes state/action/observation are torch.Tensor. Since RollingDynamicsModel (with the simulator) is a sub-class of `DynamicsModel`, I would assume that the derived class to use the same (or derived) type of the base class function arguments. Maybe this is another C++ versus python issue? That in C++, the derived class will use the parent class function argument type to implement an abstract method, but in python probably we don't enforce such rule.

**Review by blake-wulfe-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17)* status: 0 of 8 files reviewed, all discussions resolved

___
*[`mani_skill2/dynamics/base.py` line 6 at r1](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-NOuQQ0H4TYF-oeCCkPc:-NOvk7X45spdd7pWCJaE:bbsc1e7) ([raw file](https://github.com/hongkai-dai/ManiSkill2/blob/1a985edde23c45f227d2dee861f5666cfac54905/mani_skill2/dynamics/base.py#L6)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done. I changed everything to torch.Tensor

> I don't know why we would need to convert to torch.Tensor when using the gt simulator

I mean now the base class `DynamicsModel` assumes state/action/observation are torch.Tensor. Since RollingDynamicsModel (with the simulator) is a sub-class of `DynamicsModel`, I would assume that the derived class to use the same (or derived) type of the base class function arguments. Maybe this is another C++ versus python issue? That in C++, the derived class will use the parent class function argument type to implement an abstract method, but in python probably we don't enforce such rule.
</blockquote></details>

Got it. Yeah we should change this later if we start using the gt simulator but I think it works for now

**Review by blake-wulfe-tri — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/hongkai-dai/ManiSkill2/17#-:-NOvkJRx123r3-bndVH0:bnfp4nl)* status: 0 of 8 files reviewed, all discussions resolved
